brigade/core/inventory.py

```diff
@@ -189,11 +198,12 @@ def get_connection(self, connection):
             An already established connection of type ``connection``
         """
         if connection not in self.connections:
-            msg = (
-                "Couldn't find an established connection for '{c}'. "
-                "Did you call '{c}_connection'?"
-            ).format(c=connection)
-            raise AttributeError(msg)
+            try:
+                conn_task = self.brigade.available_connections[connection]
+            except KeyError:
+                raise AttributeError("not sure how to establish a connection for {}".format(
+                    connection))
+            self.brigade.filter(name=self.name).run(conn_task, num_workers=1)
```

Comment:
What is the purpose of the `filter` here?

You might want to but a  comment in the code here (since it isn't super clear from the code that conn_task and adds the connection and thus `get_connection` returns the newly created connection in the return statement. 

```
# Establishes the connection, conn_task must add connection to self.connection['connection_name']
self.brigade.filter(name=self.name).run(conn_task, num_workers=1)
```

Reply:
> What is the purpose of the filter here?

Will add a comment but the idea is that I wanted to call the connection task in the same way as a user would call it explicitly so the filter is basically saying "run this only for myself". The reason to call it like this is for two reasons:

1. To avoid having `task=None, host=None` and then `host = host or task.host` as we had before (remember that `netmiko_host` issue we had :P)
2. To avoid surprises. If we use the API in the documented way there are less chances of surprises.

> You might want to but a comment in the code here (since it isn't super clear from the code that conn_task and adds the connection and thus get_connection returns the newly created connection in the return statement.

Maybe I misinterpreted this but I think this requires two distinct documentation points:

1. We need to make very clear that `get_connection` has this behavior of establishing the connection if none exists. I will do that in the docstring of this method so users reading the docs see it. Probably a quick "how to" document would be interesting as well.
2. We need to make also very clear that if you write a connection task, the task has to set `host.connections[connetion_name]`. For that I will write a "how to write a connection task" similar to this: http://brigade.readthedocs.io/en/develop/howto/writing_a_custom_inventory.html

Reply:
Sounds good on your two points.

I also think we should just add a comment here:

```
# Establishes the connection, conn_task also adds connection to
# self.connection['connection_name']
self.brigade.filter(name=self.name).run(conn_task, num_workers=1)
```

I have looked at this code twice now (at least) and each time I have initially made the error of thinking it is not returning the connection (as it is not obvious at this point that the 'conn_task' call automatically does that).

Reply:
ok, will do